go/home/home.go

```diff
@@ -0,0 +1,205 @@
+package home
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/keybase/client/go/gregor"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/gregor1"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	"golang.org/x/net/context"
+	"sync"
+	"time"
+)
+
+type cache struct {
+	obj      keybase1.HomeScreen
+	cachedAt time.Time
+}
+
+type Home struct {
+	libkb.Contextified
+
+	sync.Mutex
+	cache *cache
+}
+
+type rawGetHome struct {
+	Status      libkb.AppStatus `json:"status"`
+	VisitRecord *struct {
+		Visits  int           `json:"visits"`
+		Atime   keybase1.Time `json:"atime"`
+		Version int           `json:"version"`
+	} `json:"visit_record"`
+	TodoList []struct {
+		Badged bool                        `json:"badged"`
+		Type   keybase1.HomeScreenTodoType `json:"type"`
+	} `json:"todo_list"`
+}
+
+func (r *rawGetHome) GetAppStatus() *libkb.AppStatus {
+	return &r.Status
+}
+
+func NewHome(g *libkb.GlobalContext) *Home {
+	home := &Home{Contextified: libkb.NewContextified(g)}
+	g.AddLogoutHook(home)
+	return home
+}
+
+func (h *Home) get(ctx context.Context, markedViewed bool) (ret keybase1.HomeScreen, err error) {
+	defer h.G().CTrace(ctx, "Home#get", func() error { return err })()
+
+	arg := libkb.NewAPIArgWithNetContext(ctx, "home")
+	arg.SessionType = libkb.APISessionTypeREQUIRED
+	arg.Args = libkb.HTTPArgs{
+		"record_visit": libkb.B{Val: markedViewed},
+	}
+	var raw rawGetHome
+	if err = h.G().API.GetDecode(arg, &raw); err != nil {
+		return ret, err
+	}
+	for _, item := range raw.TodoList {
+		ret.Items = append(ret.Items, keybase1.HomeScreenItem{
+			Data:   keybase1.NewHomeScreenItemDataWithTodo(keybase1.NewHomeScreenTodoDefault(item.Type)),
+			Badged: item.Badged,
+		})
+	}
+	if raw.VisitRecord != nil {
+		ret.LastViewed = raw.VisitRecord.Atime
+		ret.Version = raw.VisitRecord.Version
+	}
+	return ret, err
+}
+
+func (h *Home) Get(ctx context.Context, markViewed bool) (ret keybase1.HomeScreen, err error) {
+	defer h.G().CTrace(ctx, "Home#Get", func() error { return err })()
+
+	h.Lock()
+	defer h.Unlock()
+
+	if h.cache != nil {
+		diff := h.G().GetClock().Now().Sub(h.cache.cachedAt)
+		if diff < libkb.HomeCacheTimeout {
+			h.G().Log.CDebugf(ctx, "| hit cache (cached %s ago)", diff)
+			if markViewed {
+				h.G().Log.CDebugf(ctx, "| going to server to mark view, anyways")
+				tmp := h.markViewed(ctx)
```

Comment:
name `tmp` `err`

Reply:
Let's say `tmpErr` so I don't worry about it shadowing the `err` for the whole function

Reply:
Ok if you like. Shadowing wouldn't be bad here. Also you could do `err = h.markViewed(ctx)` and that would re-assign without shadowing.